OpenSim/Common/Test/testOperators.cpp

```diff
@@ -0,0 +1,421 @@
+        if (get_delay() > 0) {
+            q = _delay.getValue(s);
+        } else {
+            q = getModel().getCoordinateSet()[0].getValue(s);
+        }
+        double u = -10 * q;
+        getModel().getActuators()[0].addInControls(Vector(1, u), controls);
+    }
+    Vector getCoordValueVector(const SimTK::State& s) const {
+        return Vector(5, getInputValue<double>(s, "coord_value"));
+    }
+    // Just for testing purposes to see if the Delay itself works.
+    double getDelayedCoordinateValue(const State& s) const {
+        return _delay.getOutputValue<double>(s, "output");
+    }
+    Vector getDelayedCoordValueVector(const SimTK::State& s) const {
+        return _vectorDelay.getValue(s);
+    }
+    SimTK::Vec3 getDelayedCOMVelocityValue(const State& s) const {
+        return _comVelocityDelay.getValue(s);
+    }
+
+private:
+    void constructProperties() override {
+        constructProperty_delay(0.0);
+    }
+    void constructInputs() override {
+        constructInput<double>("coord_value", SimTK::Stage::Position);
+    }
+    void constructOutputs() override {
+        constructOutput<Vector>("coord_value_vector",
+                &PendulumController::getCoordValueVector,
+                SimTK::Stage::Position);
+    }
+    void extendFinalizeFromProperties() override {
+        Super::extendFinalizeFromProperties();
+        _delay.set_delay(get_delay());
+        _vectorDelay.set_delay(get_delay());
+        _comVelocityDelay.set_delay(get_delay());
+        addComponent(&_delay);
+        addComponent(&_vectorDelay);
+        addComponent(&_comVelocityDelay);
+    }
+    void extendConnectToModel(Model& model) override {
+        Super::extendConnectToModel(model);
+        const auto& coord = model.getCoordinateSet()[0];
+        getInput("coord_value").connect(coord.getOutput("value"));
+        _delay.getInput("input").connect(coord.getOutput("value"));
+        _vectorDelay.getInput("input").connect(getOutput("coord_value_vector"));
+        _comVelocityDelay.getInput("input").connect(
+                model.getOutput("com_velocity"));
+    }
+
+    ScalarDelay _delay;
+    VectorDelay _vectorDelay;
+    Delay<SimTK::Vec3> _comVelocityDelay;
+};
+
+/// Contains dummy outputs with different requiresAt stages so that we can
+/// test that stages are checked appropriately when asking for the Delay output.
+class DelayStageTesting : public ModelComponent {
+OpenSim_DECLARE_CONCRETE_OBJECT(DelayStageTesting, ModelComponent);
+public:
+    DelayStageTesting() {
+        constructInfrastructure();
+    }
+    // This also tests the convenience constructor.
+    ScalarDelay posDelay{0.01};
+    ScalarDelay velDelay{0.02};
+    ScalarDelay dynDelay{0.03};
+
+private:
+    void constructOutputs() override {
+        constructOutput<double>("position",
+                std::bind([](const SimTK::State& s)->double
+                        { return s.getTime() + 1.0; },
+                        std::placeholders::_1),
```

Comment:
time is a state variable, so it would be a stronger test to ask for a position level calculation result like the spatial location of a station (or one of its component to maintain the scalar output). Same for velocity and dynamics below.